### PR TITLE
DSD-234: FilterBar

### PR DESCRIPTION
**Pull request by isastettler**
Fixes JIRA ticket DSD-234

## This PR does the following:

- Adds FilterBar component
- Adds useFilterBar hook
- Adds tests for FilterBar and useFilterBar
- Adds storybook documentation for FilterBar component
- Adds storybook documentation for useFilterBar
- Updates useMultiSelect hook to work in Scout App
- Updates tests of MultiSelect and useMultiSelect accordingly
- Updates storybook for MultiSelect and useMultiSelect

## Accessibility concerns or updates

-

### Checklist:

- [x] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

### Front End Review:

- [ ] Review the Vercel preview deployment once it is ready.

### CODE REVIEW

File: src/components/ButtonGroup/ButtonGroup.tsx
Comment (EdwinGuzman) [inline src/components/ButtonGroup/ButtonGroup.tsx]: This is fine, thanks. You can remove the comment.

File: src/components/FilterBar/FilterBar.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.tsx:27]: These values, along with the ones in the `type`s, should have a comment with a description. It seems like you already have those in the Storybook doc, but it should be here as well.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.stories.mdx:72]: There should be an `Accessibility` section after `Component Props`. You can write a very brief description and say that the accessibility depends on the `FilterBar`s children. You can then link to the components that it will render, such as `MultiSelect#accessibility`, etc.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.stories.mdx:74]: Add a new heading here for the `useFilterBar hook`  section.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.stories.tsx]: Minor lint issue.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.stories.tsx]: Minor lint issue.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.stories.tsx:769]: This is is not really an issue but because every `FilterBar` uses the same props when they are rendered, then when you select a checkbox in one MultiSelect (like in Colors), then it selects the same value for all other MultiSelects.
You can keep this as is but add a sentence or two explaining that this functionality only occurs for the Storybook example.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.test.tsx]: Remove these.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.test.tsx]: Nitpick: don't import everything, do `import React from "react";`.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.test.tsx]: Nitpick: typo for `MulitSelectTestGroup`.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.test.tsx]: There should be a test for the "Apply Filters" button that appears next to the "Clear filters" button.

File: src/components/FilterBar/FilterBar.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.tsx]: Try creating a "type guard". See the `isFilterType` function here: https://github.com/NYPL/nypl-design-system/blob/development/src/components/TagSet/TagSet.tsx#L71

File: src/components/FilterBar/FilterBar.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.tsx]: This should have a different id. It's currently the same as line number 172.

File: src/hooks/useFilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/hooks/useFilterBar.stories.mdx]: typo: `modal`.

File: src/hooks/useFilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/hooks/useFilterBar.stories.mdx]: Nitpick: pass just the prop name `showClearAll`, `showSubmitAll`, etc instead of setting `={true}` which is the same.

File: src/hooks/useFilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/hooks/useFilterBar.stories.mdx:162]: I know this is an example, but it would be good to define what `multiSelectItems` is as part of the example.

File: src/hooks/useFilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/hooks/useFilterBar.stories.mdx]: typo: `structure`.

File: src/hooks/useFilterBar.stories.mdx
Comment (EdwinGuzman) [inline src/hooks/useFilterBar.stories.mdx]: This is nice but just define an example value for `initialFilter`.

File: src/hooks/useFilterBar.ts
Comment (EdwinGuzman) [inline src/hooks/useFilterBar.ts:15]: Just add a comment above the function describing it.

File: src/hooks/useMultiSelect.stories.mdx
Comment (EdwinGuzman) [inline src/hooks/useMultiSelect.stories.mdx:226]: The example in `useFilterBar.ts` does not use the `useEffect`. Make it consistent for both examples.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Works well overall! Just minor updates. Perhaps creating the type guard might be more involved.

File: src/hooks/useFilterBar.stories.mdx
Comment (isastettler) [reply src/hooks/useFilterBar.stories.mdx]: Hey @EdwinGuzman just to clarify: You suggest to remove the detailed example and just leave the data structure example code (line 199 to line 208) for the `initialFilter`? Just want to make sure I understand correctly.

Comment (EdwinGuzman) [reply src/hooks/useFilterBar.stories.mdx]: Don't remove anything. You can copy what's on line 200-207 and paste it in this example inside `FilterBarWithHookExample` or outside of it, but just to be located in the entire example you have.

File: src/components/ButtonGroup/ButtonGroup.tsx
Comment (isastettler) [reply src/components/ButtonGroup/ButtonGroup.tsx]: Removed.

File: src/components/FilterBar/FilterBar.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.tsx:27]: Added.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.mdx:72]: Added. Happy to make any edits to the text.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.mdx:74]: Done.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.tsx]: I just realized I did not address this yet. What would you like me to change?

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.tsx]: Same ^^.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.test.tsx]: Removed.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.test.tsx]: Done.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.test.tsx]: Ups. Fixed.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.test.tsx]: Added a bunch of tests...had some struggles mocking the viewport size. Now it is working but I am very aware that there might be a better way to do it. If you have an idea, pls point me in the right direction.

File: src/components/FilterBar/FilterBar.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.tsx]: I was not able to think about it in a way that would apply here. Happy to give it another try if you have a hint for me. However I added a check that my brain was able to produce.

File: src/components/FilterBar/FilterBar.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.tsx]: Fixed.

File: src/hooks/useFilterBar.stories.mdx
Comment (isastettler) [reply src/hooks/useFilterBar.stories.mdx]: Fixed.

File: src/hooks/useFilterBar.stories.mdx
Comment (isastettler) [reply src/hooks/useFilterBar.stories.mdx]: Ah yes. Done.

File: src/hooks/useFilterBar.stories.mdx
Comment (isastettler) [reply src/hooks/useFilterBar.stories.mdx:162]: To avoid adding all those lines I added it just ones on top of the page. But if you think it is better to add it to each example - or some of the examples - let me know.

File: src/hooks/useFilterBar.stories.mdx
Comment (isastettler) [reply src/hooks/useFilterBar.stories.mdx]: Fixed.

File: src/hooks/useFilterBar.stories.mdx
Comment (isastettler) [reply src/hooks/useFilterBar.stories.mdx]: Done.

File: src/hooks/useFilterBar.ts
Comment (isastettler) [reply src/hooks/useFilterBar.ts:15]: Added.

File: src/hooks/useMultiSelect.stories.mdx
Comment (isastettler) [reply src/hooks/useMultiSelect.stories.mdx:226]: Added useEffect since in Scout we have to use it it this way in order to work.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.tsx:769]: I added a fix for it. I think it's the more elegant solution ;)

File: src/components/FilterBar/FilterBar.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.tsx]: You can remove this.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (EdwinGuzman) [inline src/components/FilterBar/FilterBar.test.tsx]: You shouldn't need this because the props don't change from the initial render to the rerender. Why was this added?

**Review by EdwinGuzman — COMMENTED** [review]
The update looks good. The tests seem alright, but I'm just confused why the `rerender` is called immediately after the initial `render` even though the props don't change.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.test.tsx]: Yes so that's where I could not find a better solution. If you  add `console.log(id, isLargerThanMedium)` in FilterBar.tsx you will see that for each test the isLargerThanMedium is returning false before the rerender (exept for the first `it()` block. I tried `await`, `await waitFor()` without success, I tried to call window resize in different places and played around with the setup but this was the only thing I was able to get to work. If you have a hunch I'd love to know where to look further.

File: src/components/FilterBar/FilterBar.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.tsx]: Removed.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.tsx]: Done.

File: src/components/FilterBar/FilterBar.stories.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.tsx]: Done.

File: src/components/FilterBar/FilterBar.test.tsx
Comment (EdwinGuzman) [reply src/components/FilterBar/FilterBar.test.tsx]: Okay, there are two things going on. I tried removing the `rerender` calls and the tests do fail.

1) When I removed the `rerender` calls and only ran the test on line 232 `it("should render a 'Apply Filters' button when showSubmitAll prop is passed"...` , it passes. When _all_ tests are run, it fails. The order affects whether tests run successfully and that's not ideal. Move the "desktop" tests towards the end of the file (before the snapshot tests).
2) call `window.resizeTo(....)` _after_ the `render` function.

Try that. I managed to make all tests pass except `"should call onSubmit function when 'Apply Filters' button is clicked"`. The `mockonSubmit` function is not called so that's odd but i think that can be figured out.

Comment (isastettler) [reply src/components/FilterBar/FilterBar.test.tsx]: I changed the order of tests and was able to delete `rerender()` for all tests except the `"should call onSubmit function when 'Apply Filters' button is clicked"`. I never ran into the issue you mentioned but it kept rendering the mobile view for me.

File: src/components/FilterBar/FilterBar.tsx
Comment (bigfishdesign13) [inline src/components/FilterBar/FilterBar.tsx:82]: Is this currently the only form "group" component that is supported?

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (bigfishdesign13) [inline src/components/FilterBar/FilterBar.stories.mdx]: nit: adjust this text to the following...

...a `setSelectedItems` function to control or manipulate the (initial) state of `selectedItems`, and  a `selectedItems` object to house the current state of the selections.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (bigfishdesign13) [inline src/components/FilterBar/FilterBar.stories.mdx:126]: Add a to this section a direct link to the `useFilterBar` page under the `Hooks` heading.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (bigfishdesign13) [inline src/components/FilterBar/FilterBar.stories.mdx]: Also add info that the width of the child elements will be set to "full."

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
This is looking very good.  I added some change requests directly to the code, but the bulk of my requests are listed below.

#### Column Layout:
- The child `MultiSelect` components should be set to `width: "full"`
- The `ButtonGroup` should be set to `layout: "column"` and `buttonWidth: "full"` (similar to how the action buttons are setup for the mobile view)

#### Story Page
- In examples with where either `showClearAll` or `showSubmitAll` are `true`, the child `MultiSelect` components should be set to `variant: "listbox"`
- Update some of the `MultiSelect` components so there are enough `Checkbox` options to cause the scrollbar to kick in
- Add at least one example with two `MultiSelectGroup` components

#### Mobile View
- In the modal, the position of the heading (the title and close X) and the footer (the action buttons) should be locked in place
- In the modal, only the content area (the form elements) should scroll if the height extends beyond the available space
- Refer to this example: https://nypl.github.io/nypl-design-system/reservoir/v1/?path=/docs/components-overlays-switchers-modal--content-window-scrolling

File: src/components/FilterBar/FilterBar.tsx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.tsx:82]: yes

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.mdx]: Updated.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.mdx]: Added.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (isastettler) [reply src/components/FilterBar/FilterBar.stories.mdx:126]: I moved the link that was at  the bottom of this section to the top. Pls let me know if I misunderstood the task.

**Comment by isastettler** [comment]
> This is looking very good. I added some change requests directly to the code, but the bulk of my requests are listed below.
> 
> #### Column Layout:
> * The child `MultiSelect` components should be set to `width: "full"`
> * The `ButtonGroup` should be set to `layout: "column"` and `buttonWidth: "full"` (similar to how the action buttons are setup for the mobile view)

I have changed the default width for `layout: "column` to be` "full"` for the child components as well as the buttons and added an additional `filterWidth` prop that can control the width if needed. Bill has mentioned that on DXP we need to be able to set the width of the child components and I believe Edwin mentioned that too (I could be wrong though with Edwin). 

> 
> #### Story Page
> * In examples with where either `showClearAll` or `showSubmitAll` are `true`, the child `MultiSelect` components should be set to `variant: "listbox"`

Here I am not sure if you would like this just on storybook or in the component. In the second scenario I think it would require the Submit button to be rendered though - or do I get this wrong? 

> * Update some of the `MultiSelect` components so there are enough `Checkbox` options to cause the scrollbar to kick in
> * Add at least one example with two `MultiSelectGroup` components

Updated the examples. Let me know if this works.

> 
> #### Mobile View
> * In the modal, the position of the heading (the title and close X) and the footer (the action buttons) should be locked in place
> * In the modal, only the content area (the form elements) should scroll if the height extends beyond the available space
> * Refer to this example: https://nypl.github.io/nypl-design-system/reservoir/v1/?path=/docs/components-overlays-switchers-modal--content-window-scrolling

This should be working as expected now.

**Comment by EdwinGuzman** [comment]
The update looks good to me but I'll defer to @bigfishdesign13 on the final display review.

> Bill has mentioned that on DXP we need to be able to set the width of the child components and I believe Edwin mentioned that too (I could be wrong though with Edwin).

I don't remember this! I think if what Marty is asking for is a DS pattern and in DXP it needs to be different, it would be good to bring up why this is the case and what exactly the situation is.

File: src/components/FilterBar/FilterBar.stories.mdx
Comment (bigfishdesign13) [reply src/components/FilterBar/FilterBar.stories.mdx:126]: Oops.  I must have missed that initially.  I'm sorry.  Had the link not existed, the bottom of the section is where I would have suggested to put it.  However, I clearly misssed it when it was there, so leave it in the no position below the section heading.

File: src/components/FilterBar/FilterBar.tsx
Comment (bigfishdesign13) [reply src/components/FilterBar/FilterBar.tsx:82]: OK, that's what I initially thought.  We'll just need to add support for additional form components before other projects can use the `FilterBar` component.

**Comment by bigfishdesign13** [comment]
> Here I am not sure if you would like this just on storybook or in the component. In the second scenario I think it would require the Submit button to be rendered though - or do I get this wrong?

Sorry, I could have been a little clear.  The suggested pattern is to only use the global `Apply` and `Clear` buttons with the MultiSelect `listBox` variant.  The MultiSelect `dialog` variant has that functionality built into the component, so those global buttons are not necessary.  I would update the examples to match the suggested patterns.

**Comment by EdwinGuzman** [comment]
I haven't checked, but is that pattern in the spec doc? We're adding a bit more responsibility on the user to get the pattern right. I'm not sure how we can enforce this right now, but for the time being, would it make sense to have a section in the Storybook doc that says something like "It's suggested that the global Apply and Clear buttons be used with `listBox` MultiSelect variants." ? This can be included near the examples.

**Comment by bigfishdesign13** [comment]
> I haven't checked, but is that pattern in the spec doc? We're adding a bit more responsibility on the user to get the pattern right. I'm not sure how we can enforce this right now, but for the time being, would it make sense to have a section in the Storybook doc that says something like "It's suggested that the global Apply and Clear buttons be used with `listBox` MultiSelect variants." ? This can be included near the examples.

@EdwinGuzman 
No, this idea is somethign I neglected to include in the spec doc.  But yes, I do think it would be a good idea to include details about the expected NYPL patterns.  I don't want to hold this up, so we can create a ticket to add that documentation after the component is released.

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
This is looking better.  There are some additional styling issues that I would like to cleanup, but I will create a new ticket for those updates and I will pick it up after the release.  

There is, however, a functional bug that I found in the mobile view.  While a `MultiSelect` components is open, you can not open any `MultiSelect` components below the one that is open.  When you try to open one of the others, the screen bonces back up to the one that is already open.  If all of the `MultiSelect` components are closed, then there are no problems when opening any of the `Multiselect` components.

We can move forward to release this iteration once this bug has been resolved.

**Review by bigfishdesign13 — APPROVED** [review]
Here we go!

**Pull request merged**
